lvgl: upgrade to v8.1.0

brgl opened the pull request:
This PR contains modifications to the LVGL glue code in the kernel needed in order to support the v8 major release of LVGL.

The bulk of the changes are adjustments needed to use the LVGL interfaces that broke compatiblity with v7. There are some minor tweaks and fixes here and there too.

One important change is removing the support for using k_heap by lvgl. This was suggested to me by @npitre as implementing `k_realloc()` would be tricky and not worth the hassle as lvgl has no business using k_heap anyway. We're replacing this by allowing to use a `realloc()` implementation based directly on `sys_heap` and providing the internally implemented allocator from lvgl based on TLSF.

[Tracking issue](https://github.com/zephyrproject-rtos/zephyr/issues/40901)

This PR depends on [the following PR to the LVGL repo](https://github.com/zephyrproject-rtos/lvgl/pull/15)

Tested on `native_posix` and `nrf52840dk_nrf52840` with Adafruit 2.8" TFT display.

Closes #40901
Closes #38305

Commit:
lvgl: bump version to v8.1.0

This updates the lvgl in-tree glue code to work with version v8.1.0 and
bumps the west manifest accordingly.

The following are the most significant changes:
- The logging callback has changes in lvgl and no longer provides the
  caller with an integer log level code. We now need to parse the log
  string's prefix to determine the level.
- Several Kconfig options (mostly for default values of various settings)
  have been removed because these values are no longer configurable in
  lvgl.
- The library no longer performs a deep copy of the display and input
  device driver structs, so these must no longer be allocated on the
  stack in the init func.

Other than that it's mostly about renaming of various structures and
functions and adjusting the calls if function's signatures have changed.

This patch allows all in-tree users to work correctly but it's likely
it doesn't support all new widgets and layouts added in lvgl v8. These
however can be added gradually once this is upstream.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
samples: lvgl: add missing Kconfig options

The sample graphical Hello World program uses the image widget as well as
the Montserrat font size 14. While currently those are being selected by
default, it won't be the case once we update to lvgl v8. Select relevant
Kconfig option in prj.conf.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
lvgl: kill LVGL_MEM_POOL_HEAP_KERNEL and LVGL_MEM_POOL_KERNEL

Current lvgl code allows to use the kernel heap for dynamic memory
allocation. The k_heap API doesn't however provide k_realloc() which
will be needed in order to update lvgl to v8. Nico suggested there's
no good reason for lvgl to use k_heap and it should stick to either
the libc's allocator or depend on its own private sys_heap.

The alternative would be to extend the k_heap API to provide k_realloc()
but this may be tricky for several reasons and for now there would
be a single user anyway.

This removes the choice of using k_heap for lvgl and renames the user
pool to SYS_HEAP in Kconfig and makes it the default option.

The prj.conf for the lvgl sample is modifed to specify the number of
memory pool blocks instead of the total size as the default block
size is 2048 and it results in the same size of memory.

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
lvgl: add locking for the sys heap

The sys_heap layer doesn't provide any locking mechanism. Add a spin_lock
around the calls to sys_heap functions.

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
lvgl: provide lvgl_realloc()

In order to support lvgl v8, we need to provide a realloc() implementation
in our custom sys_heap allocator. This uses sys_heap_realloc() internally
and exposes traditional realloc() semantics.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
tests: lvgl: drop duplicate include

lvgl.h is already included in the img.h header. Drop the duplicate.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

pdgendt commented on west.yml:
```diff
@@ -157,7 +157,7 @@ manifest:
       path: modules/lib/loramac-node
     - name: lvgl
-      revision: 783c1f78c8e39751fe89d0883c8bce7336f55e94
+      revision: b63b926f0ea64d05a29d3488d615445bf98317a1
```
You should point to the PR here
```suggestion
      revision: pull/15/head
```

gmarull commented on lib/gui/lvgl/lvgl_mem.c:
```diff
@@ -16,15 +16,27 @@
 
 static char lvgl_heap_mem[HEAP_BYTES];
 static struct sys_heap lvgl_heap;
+static struct k_spinlock lvgl_heap_lock = { };
```
`= {};` redundant

carlescufi commented:
@kisvegabor @hongshui3000 @sdalu @escherstair please review

gmarull commented on lib/gui/lvgl/lvgl.c:
```diff
@@ -41,45 +46,33 @@ static uint8_t buf1[BUFFER_SIZE] __aligned(4);
-		Z_LOG(LOG_LEVEL_ERR, "%s", dupdsc);
-		break;
-	}
+	static const char *const err_str = "[Error]";
+	static const char *const wrn_str = "[Warn]";
+	static const char *const inf_str = "[Info]";
+	static const char *const dbg_str = "[Trace]";
+
+	if (strncmp(buf, err_str, strlen(err_str)) == 0)
+		LOG_ERR("%s", log_strdup(buf + strlen(err_str) + 1));
+	else if (strncmp(buf, wrn_str, strlen(wrn_str)) == 0)
+		LOG_WRN("%s", log_strdup(buf + strlen(wrn_str) + 1));
+	else if (strncmp(buf, inf_str, strlen(inf_str)) == 0)
+		LOG_INF("%s", log_strdup(buf + strlen(inf_str) + 1));
+	else if (strncmp(buf, dbg_str, strlen(dbg_str)) == 0)
+		LOG_DBG("%s", log_strdup(buf + strlen(dbg_str) + 1));
```
nit: maybe a faster way would be to just compare against `buf[1]` since it seems to be a set: `E, W, I, T`.

gmarull commented on lib/gui/lvgl/lv_conf.h:
```diff
@@ -59,7 +59,11 @@ typedef short lv_coord_t;
 
 /* Memory manager settings */
 
+#if defined(CONFIG_LVGL_MEM_POOL_TLSF)
+#define LV_MEM_CUSTOM 0
+#else
 #define LV_MEM_CUSTOM 1
+#endif
```
```c
#define LV_MEM_CUSTOM IS_ENABLED(CONFIG_LVGL_MEM_POOL_TLSF)
```

SebastianBoe replied:
( ! IS_ENABLED(...) )

gmarull reviewed (commented):
I don't see any new LVGL option being added. All LVGL options (e.g. widgets) should be exposed to Zephyr.

npitre commented:
One tiny optimization: you could add `__aligned(8)` to the lvgl_heap_mem 
array. That'll avoid wasting a few bytes in some circumstances.

brgl commented:
> Would it be possible to use lvgl's Kconfig directly? It'd make super simple to update to new minor versions .

Probably, but that would break people's existing configfs. Is that fine? (I suppose it is if we're fine with API breaking changes anyway).

carlescufi commented:
> > Would it be possible to use lvgl's Kconfig directly? It'd make super simple to update to new minor versions .
> 
> Probably, but that would break people's existing configfs. Is that fine? (I suppose it is if we're fine with API breaking changes anyway).

That would be acceptable in a case like this, yes.

Commit:
lvgl: mem: align the memory used for the private sys heap

Improve efficiency by introducing the 8-byte alignment of the private
heap.

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
lvgl: move the lvgl glue out of the zephyr tree

Start using the upstream Kconfig from LVGL and move the glue code out
of the zephyr tree and put it under lvgl/zephyr/ in modules.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

Commit:
lvgl: take opacity into account

The callback used for the 32-bit color depth has a slight problem, it
doesn't take opacity into account. Correctly mix the colors by using
lv_color_mix().

Suggested-by: Gabor Kiss-Vamosi <kisvegabor@gmail.com>
Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

dleach02 commented on lib/gui/lvgl/Kconfig.memory:
```diff
@@ -14,36 +14,24 @@ config LVGL_BITS_PER_PIXEL
 choice
 	prompt "Memory pool"
-	default LVGL_MEM_POOL_HEAP_KERNEL
+	default LVGL_MEM_POOL_SYS_HEAP
```
how come some Kconfigs are "LV_*" and others are "LVGL_*"?

brgl replied:
The LV_ come from LVGL itself. LVGL_ are options defined in Zephyr because we still need some.

dleach02 commented on samples/subsys/display/lvgl/prj.conf:
```diff
@@ -1,4 +1,4 @@
-CONFIG_HEAP_MEM_POOL_SIZE=16384
+CONFIG_LVGL_MEM_POOL_NUMBER_BLOCKS=8
```
good example of my previous comment where the Kconfigs switch preambles

brgl replied:
No, it's a different thing. The deleted line defines the global kernel heap size. The new line defines the number of heap blocks for the private LVGL heap. This is done because the whole kernel heap allocator for lvgl is now gone. The default size of a block is 2048 which results in the same pool size.

dleach02 commented:
would it make sense to normalize the Kconfig preambles to align with what LVGL if we are going to pull in the Kconfig from LVGL?

brgl commented:
> would it make sense to normalize the Kconfig preambles to align with what LVGL if we are going to pull in the Kconfig from LVGL?

I don't believe so. It's useful to distinguish the Zephyr-specific options from those coming from upstream.

MaureenHelm reviewed (dismissed):
Looks good, tested on `native_posix`.

Please review the CI documentation failure.

gmarull commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,15 +24,17 @@ config APP_LINK_WITH_LVGL
-rsource "Kconfig.text"
-rsource "Kconfig.objects"
+
+source "../modules/lib/gui/lvgl/Kconfig"
```
we can't do that: downstream projects not cloning this module will fail to build. A copy needs to be kept in Zephyr's repository (this shows that Zephyr _modules_ are not that _modular_)

brgl replied:
Done.

gmarull commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,15 +24,17 @@ config APP_LINK_WITH_LVGL
 	  disabled if the include paths for LVGL are causing aliasing
 	  issues for 'app'.
 
+config LVGL_USE_FILESYSTEM
```
I think `LV_` prefix should be used everywhere for consistency, maybe use `LV_Z` for Zephyr specific options?

brgl replied:
Done, changed to LV_Z_.

brgl commented:
> Looks good, tested on `native_posix`.
> 
> Please review the CI documentation failure.

It fails due to the issue pointed out by @gmarull. I'll need to copy lvgl's Kconfig into Zephyr unfortunately.

Commit:
lvgl: change the prefix of Zephyr-specific LVGL config options

We're now using the Kconfig copied from the upstream lvgl repository. It
uses the LV_ prefix for all options while we're using LVGL_ for
Zephyr-specific ones. Make the latter consistent with upstream but also
make sure they're distinct from lvgl's by using LV_Z_ as the prefix.

Signed-off-by: Bartosz Golaszewski <bartosz.golaszewski@huawei.com>

tejlmand commented on lib/gui/lvgl/Kconfig.upstream:
```diff
@@ -0,0 +1,867 @@
+# Kconfig file for LVGL v8.0
```
The Kconfig file for a module usually goes here:
https://github.com/zephyrproject-rtos/zephyr/tree/main/modules

which also helps to ensure the Kconfig menu for that module is only available to users when the module is available.

With this PR, the `LVGL` configuration is a available even when the LVGL module is not present.

brgl replied:
You mean I should move the entire lvgl configuration from lib/gui/lvgl/ to modules/? Or just the upstream Kconfig?

brgl replied:
@tejlmand @carlocaione is telling me that actually the entire glue code for LVGL should live in modules/ but I think this is outside the scope of this PR and should be done in a separate one. What do you think?

tejlmand replied:
> @tejlmand @carlocaione is telling me that actually the entire glue code for LVGL should live in modules/ but I think this is outside the scope of this PR and should be done in a separate one. What do you think?

I see the mess if we only mode `Kconfig.upstream` into `modules/` and not the rest of the glue code.

If you believe it to be to much to move all glue code with this PR, then please create a followup issue here:
[New Issue](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=Enhancement&template=enhancement.md&title=)

and make a link in this PR, that way this PR can be approved as we then have a followup task.

brgl replied:
Done: https://github.com/zephyrproject-rtos/zephyr/issues/41628

tejlmand commented on lib/gui/lvgl/Kconfig.upstream:
```diff
@@ -0,0 +1,867 @@
+        bool
+        default y
+
+    config LV_CONF_MINIMAL
+        bool "LVGL minimal configuration."
```
I know this file is from upstream, so in principle it shouldn't be reviewed here, but want to make a comment on this.

Looking at the code, this doesn't mean a minimal configuration which a user might expect.
Instead it means a different set of defaults which is not the same as a minimal configuration.

For example it suffers from the stuck symbol syndrome:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig

Test for yourself, for example:
```
$ cmake -GNinja -DBOARD=mimxrt1050_evk -Bbuild samples/hello_world/ -DCONFIG_LVGL=y

# Open menuconfig:
$ ninja -Cbuild menuconfig

# Navigate to: `Additional libraries → LittlevGL GUI library → LVGL configuration`
# Enable '[*] LVGL minimal configuration.'
```
and then check that the value of `LV_USE_ASSERT_NULL` is still enabled despite the fact that Minimal configuration has just been enabled.
https://github.com/zephyrproject-rtos/zephyr/blob/f89724bd78b1c2c090398f8cc7ce4f393d4f8769/lib/gui/lvgl/Kconfig.upstream#L300-L302

brgl replied:
@kisvegabor FYI ^^

tejlmand replied:
Done: https://github.com/lvgl/lvgl/issues/2976

brgl commented:
@kisvegabor [This](https://github.com/zephyrproject-rtos/zephyr/runs/4728704752) failure is due to a missing copyright in the LVGL Kconfig. What copyright is OK to add? Alternatively: @MaureenHelm can we live it as is even with this compliance check failure?

brgl commented:
@tejlmand [This](https://github.com/zephyrproject-rtos/zephyr/runs/4728704981) failure is caused by the upstream Kconfig using spaces instead of tabs. What should I do with this?

MaureenHelm commented:
> @kisvegabor [This](https://github.com/zephyrproject-rtos/zephyr/runs/4728704752) failure is due to a missing copyright in the LVGL Kconfig. What copyright is OK to add? Alternatively: @MaureenHelm can we live it as is even with this compliance check failure?

Sorry for the runaround, but I don't think we should be copying the upstream Kconfig file into the main Zephyr repo. This will be difficult to maintain and confusing to users.

@tejlmand This is the same problem I hit in https://github.com/zephyrproject-rtos/zephyr/pull/32465#issuecomment-881352308.  Can we please find a way to source the Kconfig directly from the module without having to duplicate the symbols in the main Zephyr repo? Duplicating may be ok for a handful of symbols, but it doesn't scale.

kisvegabor commented:
> This failure is due to a missing copyright in the LVGL Kconfig. What copyright is OK to add?

If you mean a copyright like [this](https://github.com/zephyrproject-rtos/zephyr/blob/b208e5811e5d357ad468af75b7ee78f565abe323/drivers/led/Kconfig#L1-L2) I can add it to the upstream LVGL Kconfig too. Anyway it should be:
```
# Copyright (c) 2022 LVGL Kft.
# SPDX-License-Identifier: MIT
```

brgl commented:
> > @kisvegabor [This](https://github.com/zephyrproject-rtos/zephyr/runs/4728704752) failure is due to a missing copyright in the LVGL Kconfig. What copyright is OK to add? Alternatively: @MaureenHelm can we live it as is even with this compliance check failure?
> 
> Sorry for the runaround, but I don't think we should be copying the upstream Kconfig file into the main Zephyr repo. This will be difficult to maintain and confusing to users.
> 
> @tejlmand This is the same problem I hit in [#32465 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/32465#issuecomment-881352308). Can we please find a way to source the Kconfig directly from the module without having to duplicate the symbols in the main Zephyr repo? Duplicating may be ok for a handful of symbols, but it doesn't scale.

Can we (same as with moving the glue code into zephyr/modules/) make it a separate task? Otherwise it'll block the merge of this PR again.

brgl commented:
I reformatted the LVGL Kconfig to follow the Zephyr guidelines + added the Copyright notice as given by @kisvegabor. Once we figure out how to include the upstream Kconfig directly, we'll decide whether to push the new formatting upstream but for now it'll make the compliance checks happy.

brgl commented:
Ugh [this](https://github.com/zephyrproject-rtos/zephyr/runs/4738380465) failure is due to the Kconfig being licensed under MIT and not Apache-2.0. OK it's clearly wrong to copy it into the zephyr repo. What do I do now?

tejlmand commented:
> @tejlmand This is the same problem I hit in [#32465 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/32465#issuecomment-881352308). Can we please find a way to source the Kconfig directly from the module without having to duplicate the symbols in the main Zephyr repo?

Correct, it is exactly the same problem.
Now the main issue here is that Zephyr Kconfig files may refer to symbols available in the upstream Kconfig but the Kconfig is not available if the project has been removed from the manifest.

And we cannot source a file from a project not available locally.

The only thing possible in such a case would be to ensure that Zephyr Kconfig code outside of `modules/<module-name>` never refers directly to upstream Kconfig symbols, all referencing to the upstream symbols are done inside the Kconfig module glue code.

This would be similar to how we do this in NCS here:
https://github.com/nrfconnect/sdk-nrf/blob/main/modules/memfault-firmware-sdk/Kconfig

and then this file which refers to upstream symbols is sourced conditionally based on availability of the module, see https://github.com/nrfconnect/sdk-nrf/pull/6080.

A solution I would be happy to upstream if it is valuable in this case.

However, that would require the LVGL symbols are not referenced outside the glue code, or limited referenced so that we can handle symbol duplication.

> Duplicating may be ok for a handful of symbols, but it doesn't scale.

Exactly, which is why referencing to symbols placed in external modules should be done only in module glue code, when the Kconfig is placed external.

And thus the reason why I accepted a pure copy of the upstream Kconfig in this case.
Plust the fact that the upstream [module.yml](https://github.com/lvgl/lvgl/blob/master/zephyr/module.yml) is missing a Kconfig entry like this:
```
build:
  kconfig: .
```

tejlmand commented:
> What do I do now?

I guess we need to source it from the remote repo directly, that means this file [zephyr/module.yml](https://github.com/lvgl/lvgl/blob/master/zephyr/module.yml) must have the following content:
```
build:
  kconfig: .
```

and we must ensure that we do not directly refer to upstream settings when that module is not available, see more here: https://github.com/zephyrproject-rtos/zephyr/pull/41336#issuecomment-1007466555

tejlmand commented:
> Ugh this failure is due to the Kconfig being licensed under MIT and not Apache-2.0. OK it's clearly wrong to copy it into the zephyr repo. What do I do now?

MIT is a permissive license, so you can also try to get approval, which should be acceptable here, see more:
https://docs.zephyrproject.org/latest/contribute/external.html

brgl commented:
> > What do I do now?
> 
> I guess we need to source it from the remote repo directly, that means this file [zephyr/module.yml](https://github.com/lvgl/lvgl/blob/master/zephyr/module.yml) must have the following content:
> 
> ```
> build:
>   kconfig: .
> ```
> 
> and we must ensure that we do not directly refer to upstream settings when that module is not available, see more here: [#41336 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/41336#issuecomment-1007466555)

It's not entirely clear whether "referencing" means from C code only or from defconfigs and config fragments too? There are no references to LVGL Kconfig symbols from C code outside lib/gui/lvgl/ but there are a lot in various board and sample config files. It would be entirely impractible to duplicate those.

tejlmand commented:
> It's not entirely clear whether "referencing" means from C code only or from defconfigs and config fragments too?

This means referencing from other Kconfig files.

For example:
```
config FOO
    bool "boolean"
    depends on BAR
    selects BAZ
```

requires `BAR` and `BAZ` to be known downstream.
If `FOO` is defined on Kconfig glue code, things can work with the solution I referred https://github.com/zephyrproject-rtos/zephyr/pull/41336#issuecomment-1007466555.

If `FOO` is defined, for example here [Kconfig.zephyr](https://github.com/zephyrproject-rtos/zephyr/blob/main/Kconfig.zephyr), then we would need dummy `BAR` and `BAZ` when upstream is not available, for example:
```
if !ZEPHYR_LVGL_MODULE

# dummy symbols to avoid Kconfig failures when lvgl is unavailable
config BAR
    bool

config BAZ
    bool
 
endif
```

brgl commented:
> > It's not entirely clear whether "referencing" means from C code only or from defconfigs and config fragments too?
> 
> This means referencing from other Kconfig files.
> 
> For example:
> 
> ```
> config FOO
>     bool "boolean"
>     depends on BAR
>     selects BAZ
> ```
> 
> requires `BAR` and `BAZ` to be known downstream. If `FOO` is defined on Kconfig glue code, things can work with the solution I referred [#41336 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/41336#issuecomment-1007466555).
> 
> If `FOO` is defined, for example here [Kconfig.zephyr](https://github.com/zephyrproject-rtos/zephyr/blob/main/Kconfig.zephyr), then we would need dummy `BAR` and `BAZ` when upstream is not available, for example:
> 
> ```
> if !ZEPHYR_LVGL_MODULE
> 
> # dummy hidden symbols to avoid Kconfig failures when lvgl is unavailable
> config BAR
>     bool
> 
> config BAZ
>     bool
>  
> endif
> ```

Looks like this is all that is so it should be fine. I thought there would be more to mock.

```
./boards/shields/buydisplay_3_5_tft_touch_arduino/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/buydisplay_3_5_tft_touch_arduino/Kconfig.defconfig:	default LV_COLOR_DEPTH_16
./boards/shields/ssd1306/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/ssd1306/Kconfig.defconfig:	default LV_COLOR_DEPTH_1
./boards/shields/waveshare_epaper/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/waveshare_epaper/Kconfig.defconfig:	default LV_COLOR_DEPTH_1
./boards/shields/ls0xx_generic/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/ls0xx_generic/Kconfig.defconfig:	default LV_COLOR_DEPTH_1
./boards/shields/buydisplay_2_8_tft_touch_arduino/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/buydisplay_2_8_tft_touch_arduino/Kconfig.defconfig:	default LV_COLOR_DEPTH_16
./boards/shields/buydisplay_2_8_tft_touch_arduino/Kconfig.defconfig:config LV_COLOR_16_SWAP
./boards/shields/st7789v_generic/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/st7789v_generic/Kconfig.defconfig:	default LV_COLOR_DEPTH_16 if SHIELD_ST7789V_WAVESHARE_240X240
./boards/shields/st7789v_generic/Kconfig.defconfig:config LV_COLOR_16_SWAP
./boards/shields/adafruit_2_8_tft_touch_v2/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/shields/adafruit_2_8_tft_touch_v2/Kconfig.defconfig:	default LV_COLOR_DEPTH_32
./boards/arm/mimxrt1060_evk/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/arm/mimxrt1060_evk/Kconfig.defconfig:	default LV_COLOR_DEPTH_16
./boards/arm/mimxrt1050_evk/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/arm/mimxrt1050_evk/Kconfig.defconfig:	default LV_COLOR_DEPTH_16
./boards/arm/reel_board/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/arm/reel_board/Kconfig.defconfig:	default LV_COLOR_DEPTH_1
./boards/arm/mimxrt1064_evk/Kconfig.defconfig:choice LV_COLOR_DEPTH_BITS
./boards/arm/mimxrt1064_evk/Kconfig.defconfig:	default LV_COLOR_DEPTH_16
```

brgl commented:
All the compliance checks now pass with the external Kconfig but the documentation checks still fail. It looks like it's unrelated to the code changes. Any advice?

brgl commented:
> We have removed `HOR/VER_RES_MAX` from LVGL from v8. Now the resolution is not limited and can be set to any value when initializing the display driver.

Thanks, we use it internally in the glue code. I will take a look at removing it once this is upstream.

brgl commented:
@gmarull @tejlmand  Any comments? Can we get this merged in this state?

brgl commented:
https://github.com/zephyrproject-rtos/zephyr/runs/4772574868?check_suite_focus=true This is triggered by the build which selects LVGL but doesn't have the external lvgl module. I'm not sure why this would happen though.

gmarull reviewed (commented):
I've just tried this PR on a [downstream project](https://github.com/teslabs/spinner) that doesn't fetch LVGL, and it fails to build (because of LVGL import, same as docs build)

```diff
diff --git a/west.yml b/west.yml
index fc307f1..616246d 100644
--- a/west.yml
+++ b/west.yml
@@ -13,7 +13,7 @@ manifest:
     - name: zephyr
       remote: zephyr
       repo-path: zephyr
-      revision: main
+      revision: pull/41336/head
       import:
         name-allowlist:
           - cmsis
```

I think that to make things simple and get this in soon, we should keep a Kconfig copy in Zephyr (and adjust it as necessary to make it compliant). Then, we should figure out a better way to solve this.

I think that this PR has shown that the way modules work/are integrated needs to be improved. There are certain dependencies like for instance LVGL, canopennode, etc that are not required by Zephyr core functionality. I think it would be better to think about them as _packages_ that [applications](https://github.com/zephyrproject-rtos/example-application) pull as required, same as most modern languages do (`pip`, `cargo`, etc.). I suppose they are now in-tree to make things easier, even though it comes with associated problems. For example, if as an application developer, I need a BLE feature in Zephyr 3.x, but I need to keep the UI on LVGL 7.x I'll be in trouble, even if 7.x could perfectly run on Zephyr 3.x. And don't tell people that they need to migrate their few thousand-line UI code because of a BLE feature that is completely independent: it doesn't make sense. The other problem is: what qualifies to go under `modules/`? Will other GUI libs be accepted if they ever come? Does that scale? _All these points can be discussed out of this PR, just wanted to make a note_ :-)

brgl commented:
@kisvegabor Any chance of you relicensing/double-licensing the Kconfig under Apache-2.0 to avoid the trouble of getting the MIT upstream?

gmarull reviewed (dismissed):
LGTM, tested on `nrf52840dk_nrf52840` + `buydisplay_2_8_tft_touch_arduino`.

brgl commented:
https://github.com/zephyrproject-rtos/zephyr/issues/41804 to track the inclusion of the MIT code from this PR. @nashif please take a look as it is my understanding that you're the only person who can merge this if the CI is not green.

kisvegabor commented:
> Any chance of you relicensing/double-licensing the Kconfig under Apache-2.0 to avoid the trouble of getting the MIT upstream?

It's a little bit hard for me to foresee the legal consequences of having a dual-licensed file in an "plain" MIT repo. 
AFAIK, dual licensing means people can choose the license which fits better. Can you confirm it? If so, I see no problem, as one of the licenses is still MIT.  
@embeddedt can you comment on this?

embeddedt commented:
That's my understanding of dual-licensing as well, but I'm not a lawyer and it's not a subject I've done enough research on to give a confident answer.

brgl commented:
The version I pushed now (for this PR and the companion lvgl PR) moves the lvgl glue code out of lib/gui/lvgl/ into modules/lib/gui/lvgl/zephyr/ as well as moves the configuration into zephyr/modules/ as discussed during the dev-review meeting.

brgl commented:
Pushed a new version rebased on top of current main.

brgl commented:
Am I looking at the right thing in CI?

```
-- Found BOARD.dts: /__w/zephyr/zephyr/boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts
devicetree error: 'zephyr,memory-region' appears in /sdram@d0000000 in /__w/zephyr/zephyr/twister-out/stm32h7b3i_dk/tests/kernel/common/kernel.common/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in /__w/zephyr/zephyr/dts/bindings/sram/mmio-sram.yaml
CMake Error at /__w/zephyr/zephyr/cmake/dts.cmake:205 (message):
  gen_defines.py failed with return code: 1
```

This doesn't look like an error introduced by this PR.

gmarull commented:
> Am I looking at the right thing in CI?
> 
> ```
> -- Found BOARD.dts: /__w/zephyr/zephyr/boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts
> devicetree error: 'zephyr,memory-region' appears in /sdram@d0000000 in /__w/zephyr/zephyr/twister-out/stm32h7b3i_dk/tests/kernel/common/kernel.common/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in /__w/zephyr/zephyr/dts/bindings/sram/mmio-sram.yaml
> CMake Error at /__w/zephyr/zephyr/cmake/dts.cmake:205 (message):
>   gen_defines.py failed with return code: 1
> ```
> 
> This doesn't look like an error introduced by this PR.

rebase, should have been fixed.

jfischer-no commented:
Looks like the trend to consume more flash (almost doubled) continues. Is it okay, does it bother anyone?

**samples/subsys/display/lvgl #41336 (v8.1.0)**
                                                          
```
Memory region         Used Size  Region Size  %age Used                                                                                      
           FLASH:      191480 B         1 MB     18.26%                                                                                      
            SRAM:       26096 B       256 KB      9.95%                                                                                      
        IDT_LIST:          0 GB         2 KB      0.00%                                                                                      
Generating files from zephyr.elf for board: reel_board       
```                                                                                
                                                          
```
Memory region         Used Size  Region Size  %age Used                                                                                      
           OCRAM:          0 GB       256 KB      0.00%                                                                                      
           FLASH:      200028 B        64 MB      0.30%                                                                                      
            SRAM:      634368 B        32 MB      1.89%                                                                                      
            ITCM:          0 GB       128 KB      0.00%                                                                                      
            DTCM:          0 GB       128 KB      0.00%                                                                                      
        IDT_LIST:          0 GB         2 KB      0.00%                                                                                      
Generating files from zephyr.elf for board: mimxrt1050_evk
``` 
Unnecessary options disabled:

CONFIG_LV_CONF_MINIMAL=y
CONFIG_LV_USE_LOG is not set
CONFIG_LV_USE_TABLE is not set
Extra Widgets disabled
CONFIG_LV_USE_THEME_BASIC is not set
CONFIG_LV_USE_SNAPSHOT is not set
CONFIG_LV_USE_SNAPSHOT is not set

```
Memory region         Used Size  Region Size  %age Used                                                                                      
           OCRAM:          0 GB       256 KB      0.00%                                                                                      
           FLASH:      160528 B        64 MB      0.24%                                                                                      
            SRAM:      634272 B        32 MB      1.89%                                                                                      
            ITCM:          0 GB       128 KB      0.00%                                                                                      
            DTCM:          0 GB       128 KB      0.00%                                                                                      
        IDT_LIST:          0 GB         2 KB      0.00%                                                                                      
Generating files from zephyr.elf for board: mimxrt1050_evk    
```                                                                               
---
**samples/subsys/display/lvgl main**:  
                                                           
```
Memory region         Used Size  Region Size  %age Used                                                                                      
           FLASH:       90528 B         1 MB      8.63%                                                                                      
            SRAM:       25712 B       256 KB      9.81%                                                                                      
        IDT_LIST:          0 GB         2 KB      0.00%                                                                                      
Generating files from zephyr.elf for board: reel_board 
```

```
Memory region         Used Size  Region Size  %age Used                                                                                      
           OCRAM:          0 GB       256 KB      0.00%                                                                                      
           FLASH:       99360 B        64 MB      0.15%                                                                                      
            SRAM:      633952 B        32 MB      1.89%                                                                                      
            ITCM:          0 GB       128 KB      0.00%                                                                                      
            DTCM:          0 GB       128 KB      0.00%                                                                                      
        IDT_LIST:          0 GB         2 KB      0.00%                                                                                      
Generating files from zephyr.elf for board: mimxrt1050_evk 
```

gmarull commented:
> Looks like the trend to consume more flash (almost doubled) continues. Is it okay, does it bother anyone?
> 

I remember we had the same discussion when upgrading to v7. Maybe LVGL has decided that their main target are high-end MCUS, where this consumption may be considered _okay_ for the features you get. I personally think Zephyr shouldn't care at all, unless there's an obvious issue. If people don't like LVGL because they consider it's resource hungry, they can choose another GUI lib.

jfischer-no commented:
> > Looks like the trend to consume more flash (almost doubled) continues. Is it okay, does it bother anyone?
> 
> I remember we had the same discussion when upgrading to v7. Maybe LVGL has decided that their main target are high-end MCUS, where this consumption may be considered _okay_ for the features you get. I personally think Zephyr shouldn't care at all, unless there's an obvious issue. If people don't like LVGL because they consider it's resource hungry, they can choose another GUI lib.

That is exactly what I was thinking. Should it then be provided by Zephyr project at all? Can it be an external module without any dependencies in Zephyr?

brgl commented:
> > > Looks like the trend to consume more flash (almost doubled) continues. Is it okay, does it bother anyone?
> > 
> > 
> > I remember we had the same discussion when upgrading to v7. Maybe LVGL has decided that their main target are high-end MCUS, where this consumption may be considered _okay_ for the features you get. I personally think Zephyr shouldn't care at all, unless there's an obvious issue. If people don't like LVGL because they consider it's resource hungry, they can choose another GUI lib.
> 
> That is exactly what I was thinking. Should it then be provided by Zephyr project at all? Can it be an external module without any dependencies in Zephyr?

It mostly is with this series. The glue code is moved to the lvgl module. We still have to keep the defconfig etc. but apart from samples there's no lvgl code in the tree anymore. I just kept the history of the changes in the zephyr tree to make the review easier and keep the history in git.

kisvegabor commented:
There shouldn't be such a huge increase in flash usage. 

Please run these on an elf file:
```
  nm -S --size-sort program.elf -t d | grep -P "^[0-9]+ [0-9]+ [tT]" | tail -50

  nm -S --size-sort program.elf -t d | grep -P "^[0-9]+ [0-9]+ [rR]" | tail -50
```

brgl commented:
The changes to `lvgl` module have been merged. I rebased this PR on top of current `main` and updated the west revision.

kisvegabor commented:
@jfischer-no

You can also attach one of the elf files and I'll investigate the cause of the bloat.
I also want to confirm that LVGL still targets low end devices too and claims to fit into 64 kB flash. According to my not so old measurements it's still true but I opened [an issue in LVGL](https://github.com/lvgl/lvgl/issues/3127) to ad CI tests for flash an RAM usage.
